### PR TITLE
make on_lines function calls async

### PR DESCRIPTION
**Pull request by sunjon**

### CODE REVIEW

**Comment by sunjon** [comment]
Can anyone watching https://github.com/p00f/nvim-ts-rainbow/issues/5 test please?

**Comment by p00f** [comment]
too async?

**Comment by p00f** [comment]
it doesn't colour new brackets

File: lua/rainbow/internal.lua [hunk 1]
```diff
@@ -40,8 +47,19 @@ end
 
 local M = {}
 
+local function try_async(f)
+  if not f then return end
```
Comment (sunjon) [inline lua/rainbow/internal.lua:51]: I put in this line to stop the `f is nil` errors.
I think resolving why it's nil is related to why there's no updates after attach.

Comment (p00f) [reply lua/rainbow/internal.lua:51]: i thought of that too. yay i'm thinking right!

Comment (p00f) [reply lua/rainbow/internal.lua:51]: i'll make callbackfn(bufnr) return 0 or somehting?

Comment (p00f) [reply lua/rainbow/internal.lua:51]: nah

Comment (sunjon) [reply lua/rainbow/internal.lua]: think I've fixed it

File: lua/rainbow/internal.lua [hunk 2]
```diff
@@ -68,11 +69,11 @@ function M.attach(bufnr, lang)
     return
   end
 
-  callbackfn(bufnr) -- do it on intial load
+  callbackfn(bufnr) -- do it on attach
   vim.api.nvim_buf_attach( --do it on every change
     bufnr,
     false,
-    { on_lines = try_async(callbackfn(bufnr))}
+    { on_lines = try_async(callbackfn)}
```
Comment (p00f) [inline lua/rainbow/internal.lua:76]: why did you remove the bufnr argument?

Comment (p00f) [reply lua/rainbow/internal.lua:76]: changing it to
```lua
 on_lines = try_async(callbackfn, bufnr)}
```
slows it down again

**Review by p00f — COMMENTED** [review]
`Error executing vim.schedule lua callback: /usr/share/nvim/runtime/lua/vim/treesitter/query.lua:341: bad argument #2 to 'handler' number expected, got nil)`

line in question - https://github.com/neovim/neovim/blob/adbf2dd86b8c7d898843bdf47b89b423f62bee4b/runtime/lua/vim/treesitter/query.lua#L315

its second argument is `pattern`, why would that be a number?


getting nil is another question

**Comment by sunjon** [comment]
Let me think about this and come back to you shortly.

**Comment by p00f** [comment]
My "tests" for slowdown were typing random stuff.
There is no slowdown when typing random stuff wrapped inside a string.
Maybe you did really fix it and it hangs because of treesitter errors (wrong syntax)?

**Comment by p00f** [comment]
image
nvm the random stuff is a valid node

**Comment by p00f** [comment]
ok the above comment was for clojure, the slowdown still happens inside a string in cpp

**Comment by p00f** [comment]
I just tested one case where there was still a lag but significantly reduced with this PR

change it to
```lua
​ on_lines ​=​ ​try_async​(callbackfn, bufnr)}
```
 and I'll merge this

**Comment by p00f** [comment]
Thank you so much!!! :D

File: lua/rainbow/internal.lua (hunk 2 above)
Comment (p00f) [reply lua/rainbow/internal.lua:76]: @sunjon just to be clear, was this removal intentional? `try_async` was defined with 2 args and adding the 2nd `bufnr` argument fixes the error
